diff --git a/gel-core/policies.d.cts b/gel-core/policies.d.cts
index 794daf4..00d4263 100644
--- a/gel-core/policies.d.cts
+++ b/gel-core/policies.d.cts
@@ -10,7 +10,10 @@ export interface GelPolicyConfig {
     using?: SQL;
     withCheck?: SQL;
 }
-export declare class GelPolicy implements GelPolicyConfig {
+type GelPolicyState = {
+    [K in keyof GelPolicyConfig]-?: GelPolicyConfig[K] | undefined;
+};
+export declare class GelPolicy implements GelPolicyState {
     readonly name: string;
     static readonly [entityKind]: string;
     readonly as: GelPolicyConfig['as'];
diff --git a/gel-core/policies.d.ts b/gel-core/policies.d.ts
index 7687c59..387fb9f 100644
--- a/gel-core/policies.d.ts
+++ b/gel-core/policies.d.ts
@@ -10,7 +10,10 @@ export interface GelPolicyConfig {
     using?: SQL;
     withCheck?: SQL;
 }
-export declare class GelPolicy implements GelPolicyConfig {
+type GelPolicyState = {
+    [K in keyof GelPolicyConfig]-?: GelPolicyConfig[K] | undefined;
+};
+export declare class GelPolicy implements GelPolicyState {
     readonly name: string;
     static readonly [entityKind]: string;
     readonly as: GelPolicyConfig['as'];
diff --git a/gel-core/query-builders/query.d.cts b/gel-core/query-builders/query.d.cts
index 466729b..bad731c 100644
--- a/gel-core/query-builders/query.d.cts
+++ b/gel-core/query-builders/query.d.cts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.cjs";
 import { QueryPromise } from "../../query-promise.cjs";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.cjs";
 import type { RunnableQuery } from "../../runnable-query.cjs";
-import type { Query, SQLWrapper } from "../../sql/sql.cjs";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.cjs";
 import type { KnownKeysOnly } from "../../utils.cjs";
 import type { GelDialect } from "../dialect.cjs";
 import type { GelPreparedQuery, GelSession, PreparedQueryConfig } from "../session.cjs";
@@ -40,6 +40,7 @@ export declare class GelRelationalQuery<TResult> extends QueryPromise<TResult> i
         execute: TResult;
     }>;
     private _getQuery;
+    getSQL(): SQL;
     private _toSQL;
     toSQL(): Query;
     execute(): Promise<TResult>;
diff --git a/gel-core/query-builders/query.d.ts b/gel-core/query-builders/query.d.ts
index 48db533..899f570 100644
--- a/gel-core/query-builders/query.d.ts
+++ b/gel-core/query-builders/query.d.ts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.js";
 import { QueryPromise } from "../../query-promise.js";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.js";
 import type { RunnableQuery } from "../../runnable-query.js";
-import type { Query, SQLWrapper } from "../../sql/sql.js";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.js";
 import type { KnownKeysOnly } from "../../utils.js";
 import type { GelDialect } from "../dialect.js";
 import type { GelPreparedQuery, GelSession, PreparedQueryConfig } from "../session.js";
@@ -40,6 +40,7 @@ export declare class GelRelationalQuery<TResult> extends QueryPromise<TResult> i
         execute: TResult;
     }>;
     private _getQuery;
+    getSQL(): SQL;
     private _toSQL;
     toSQL(): Query;
     execute(): Promise<TResult>;
diff --git a/gel-core/roles.d.cts b/gel-core/roles.d.cts
index cd5e0d5..38e9a48 100644
--- a/gel-core/roles.d.cts
+++ b/gel-core/roles.d.cts
@@ -4,8 +4,14 @@ export interface GelRoleConfig {
     createRole?: boolean;
     inherit?: boolean;
 }
-export declare class GelRole implements GelRoleConfig {
+type GelRoleState = {
+    [K in keyof GelRoleConfig]-?: GelRoleConfig[K] | undefined;
+};
+export declare class GelRole implements GelRoleState {
     readonly name: string;
+    createDb: GelRoleConfig['createDb'];
+    createRole: GelRoleConfig['createRole'];
+    inherit: GelRoleConfig['inherit'];
     static readonly [entityKind]: string;
     constructor(name: string, config?: GelRoleConfig);
     existing(): this;
diff --git a/gel-core/roles.d.ts b/gel-core/roles.d.ts
index fcb1dd2..07d54fc 100644
--- a/gel-core/roles.d.ts
+++ b/gel-core/roles.d.ts
@@ -4,8 +4,14 @@ export interface GelRoleConfig {
     createRole?: boolean;
     inherit?: boolean;
 }
-export declare class GelRole implements GelRoleConfig {
+type GelRoleState = {
+    [K in keyof GelRoleConfig]-?: GelRoleConfig[K] | undefined;
+};
+export declare class GelRole implements GelRoleState {
     readonly name: string;
+    createDb: GelRoleConfig['createDb'];
+    createRole: GelRoleConfig['createRole'];
+    inherit: GelRoleConfig['inherit'];
     static readonly [entityKind]: string;
     constructor(name: string, config?: GelRoleConfig);
     existing(): this;
diff --git a/mysql-core/query-builders/delete.d.cts b/mysql-core/query-builders/delete.d.cts
index 20e968b..a542b5f 100644
--- a/mysql-core/query-builders/delete.d.cts
+++ b/mysql-core/query-builders/delete.d.cts
@@ -73,6 +73,7 @@ export declare class MySqlDeleteBase<TTable extends MySqlTable, TQueryResult ext
     orderBy(builder: (deleteTable: TTable) => ValueOrArray<MySqlColumn | SQL | SQL.Aliased>): MySqlDeleteWithout<this, TDynamic, 'orderBy'>;
     orderBy(...columns: (MySqlColumn | SQL | SQL.Aliased)[]): MySqlDeleteWithout<this, TDynamic, 'orderBy'>;
     limit(limit: number | Placeholder): MySqlDeleteWithout<this, TDynamic, 'limit'>;
+    getSQL(): SQL;
     toSQL(): Query;
     prepare(): MySqlDeletePrepare<this>;
     execute: ReturnType<this['prepare']>['execute'];
diff --git a/mysql-core/query-builders/delete.d.ts b/mysql-core/query-builders/delete.d.ts
index 9149eb1..0a71ebe 100644
--- a/mysql-core/query-builders/delete.d.ts
+++ b/mysql-core/query-builders/delete.d.ts
@@ -73,6 +73,7 @@ export declare class MySqlDeleteBase<TTable extends MySqlTable, TQueryResult ext
     orderBy(builder: (deleteTable: TTable) => ValueOrArray<MySqlColumn | SQL | SQL.Aliased>): MySqlDeleteWithout<this, TDynamic, 'orderBy'>;
     orderBy(...columns: (MySqlColumn | SQL | SQL.Aliased)[]): MySqlDeleteWithout<this, TDynamic, 'orderBy'>;
     limit(limit: number | Placeholder): MySqlDeleteWithout<this, TDynamic, 'limit'>;
+    getSQL(): SQL;
     toSQL(): Query;
     prepare(): MySqlDeletePrepare<this>;
     execute: ReturnType<this['prepare']>['execute'];
diff --git a/mysql-core/query-builders/select.d.cts b/mysql-core/query-builders/select.d.cts
index f04f97c..84f8f4a 100644
--- a/mysql-core/query-builders/select.d.cts
+++ b/mysql-core/query-builders/select.d.cts
@@ -597,6 +597,7 @@ export declare abstract class MySqlSelectQueryBuilderBase<THKT extends MySqlSele
      * @param config the lock configuration.
      */
     for(strength: LockStrength, config?: LockConfig): MySqlSelectWithout<this, TDynamic, 'for'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): MySqlSelectDynamic<this>;
diff --git a/mysql-core/query-builders/select.d.ts b/mysql-core/query-builders/select.d.ts
index d435ba4..3ebeef4 100644
--- a/mysql-core/query-builders/select.d.ts
+++ b/mysql-core/query-builders/select.d.ts
@@ -597,6 +597,7 @@ export declare abstract class MySqlSelectQueryBuilderBase<THKT extends MySqlSele
      * @param config the lock configuration.
      */
     for(strength: LockStrength, config?: LockConfig): MySqlSelectWithout<this, TDynamic, 'for'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): MySqlSelectDynamic<this>;
diff --git a/mysql-core/query-builders/select.types.d.cts b/mysql-core/query-builders/select.types.d.cts
index 8d996e7..41d84f1 100644
--- a/mysql-core/query-builders/select.types.d.cts
+++ b/mysql-core/query-builders/select.types.d.cts
@@ -102,7 +102,7 @@ export interface MySqlSelectHKT extends MySqlSelectHKTBase {
     _type: MySqlSelectBase<this['tableName'], Assume<this['selection'], ColumnsSelection>, this['selectMode'], Assume<this['preparedQueryHKT'], PreparedQueryHKTBase>, Assume<this['nullabilityMap'], Record<string, JoinNullability>>, this['dynamic'], this['excludedMethods'], Assume<this['result'], any[]>, Assume<this['selectedFields'], ColumnsSelection>>;
 }
 export type MySqlSetOperatorExcludedMethods = 'where' | 'having' | 'groupBy' | 'session' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'for';
-export type MySqlSelectWithout<T extends AnyMySqlSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<MySqlSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type MySqlSelectWithout<T extends AnyMySqlSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | MySqlSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<MySqlSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type MySqlSelectPrepare<T extends AnyMySqlSelect> = PreparedQueryKind<T['_']['preparedQueryHKT'], MySqlPreparedQueryConfig & {
     execute: T['_']['result'];
     iterator: T['_']['result'][number];
diff --git a/mysql-core/query-builders/select.types.d.ts b/mysql-core/query-builders/select.types.d.ts
index 5b2e701..d6a60a0 100644
--- a/mysql-core/query-builders/select.types.d.ts
+++ b/mysql-core/query-builders/select.types.d.ts
@@ -102,7 +102,7 @@ export interface MySqlSelectHKT extends MySqlSelectHKTBase {
     _type: MySqlSelectBase<this['tableName'], Assume<this['selection'], ColumnsSelection>, this['selectMode'], Assume<this['preparedQueryHKT'], PreparedQueryHKTBase>, Assume<this['nullabilityMap'], Record<string, JoinNullability>>, this['dynamic'], this['excludedMethods'], Assume<this['result'], any[]>, Assume<this['selectedFields'], ColumnsSelection>>;
 }
 export type MySqlSetOperatorExcludedMethods = 'where' | 'having' | 'groupBy' | 'session' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'for';
-export type MySqlSelectWithout<T extends AnyMySqlSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<MySqlSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type MySqlSelectWithout<T extends AnyMySqlSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | MySqlSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<MySqlSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type MySqlSelectPrepare<T extends AnyMySqlSelect> = PreparedQueryKind<T['_']['preparedQueryHKT'], MySqlPreparedQueryConfig & {
     execute: T['_']['result'];
     iterator: T['_']['result'][number];
diff --git a/pg-core/policies.d.cts b/pg-core/policies.d.cts
index fe8b4ee..d1bb24d 100644
--- a/pg-core/policies.d.cts
+++ b/pg-core/policies.d.cts
@@ -10,7 +10,10 @@ export interface PgPolicyConfig {
     using?: SQL;
     withCheck?: SQL;
 }
-export declare class PgPolicy implements PgPolicyConfig {
+type PgPolicyState = {
+    [K in keyof PgPolicyConfig]-?: PgPolicyConfig[K] | undefined;
+};
+export declare class PgPolicy implements PgPolicyState {
     readonly name: string;
     static readonly [entityKind]: string;
     readonly as: PgPolicyConfig['as'];
diff --git a/pg-core/policies.d.ts b/pg-core/policies.d.ts
index 9f4339c..3e5aee7 100644
--- a/pg-core/policies.d.ts
+++ b/pg-core/policies.d.ts
@@ -10,7 +10,10 @@ export interface PgPolicyConfig {
     using?: SQL;
     withCheck?: SQL;
 }
-export declare class PgPolicy implements PgPolicyConfig {
+type PgPolicyState = {
+    [K in keyof PgPolicyConfig]-?: PgPolicyConfig[K] | undefined;
+};
+export declare class PgPolicy implements PgPolicyState {
     readonly name: string;
     static readonly [entityKind]: string;
     readonly as: PgPolicyConfig['as'];
diff --git a/pg-core/query-builders/query.d.cts b/pg-core/query-builders/query.d.cts
index 566e120..1dc39a0 100644
--- a/pg-core/query-builders/query.d.cts
+++ b/pg-core/query-builders/query.d.cts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.cjs";
 import { QueryPromise } from "../../query-promise.cjs";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.cjs";
 import type { RunnableQuery } from "../../runnable-query.cjs";
-import type { Query, SQLWrapper } from "../../sql/sql.cjs";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.cjs";
 import type { KnownKeysOnly } from "../../utils.cjs";
 import type { PgDialect } from "../dialect.cjs";
 import type { PgPreparedQuery, PgSession, PreparedQueryConfig } from "../session.cjs";
@@ -40,6 +40,7 @@ export declare class PgRelationalQuery<TResult> extends QueryPromise<TResult> im
         execute: TResult;
     }>;
     private _getQuery;
+    getSQL(): SQL;
     private _toSQL;
     toSQL(): Query;
     private authToken?;
diff --git a/pg-core/query-builders/query.d.ts b/pg-core/query-builders/query.d.ts
index 94e995b..6591a68 100644
--- a/pg-core/query-builders/query.d.ts
+++ b/pg-core/query-builders/query.d.ts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.js";
 import { QueryPromise } from "../../query-promise.js";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.js";
 import type { RunnableQuery } from "../../runnable-query.js";
-import type { Query, SQLWrapper } from "../../sql/sql.js";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.js";
 import type { KnownKeysOnly } from "../../utils.js";
 import type { PgDialect } from "../dialect.js";
 import type { PgPreparedQuery, PgSession, PreparedQueryConfig } from "../session.js";
@@ -40,6 +40,7 @@ export declare class PgRelationalQuery<TResult> extends QueryPromise<TResult> im
         execute: TResult;
     }>;
     private _getQuery;
+    getSQL(): SQL;
     private _toSQL;
     toSQL(): Query;
     private authToken?;
diff --git a/pg-core/roles.d.cts b/pg-core/roles.d.cts
index 65e7369..44f352e 100644
--- a/pg-core/roles.d.cts
+++ b/pg-core/roles.d.cts
@@ -4,8 +4,14 @@ export interface PgRoleConfig {
     createRole?: boolean;
     inherit?: boolean;
 }
-export declare class PgRole implements PgRoleConfig {
+type PgRoleState = {
+    [K in keyof PgRoleConfig]-?: PgRoleConfig[K] | undefined;
+};
+export declare class PgRole implements PgRoleState {
     readonly name: string;
+    createDb: PgRoleConfig['createDb'];
+    createRole: PgRoleConfig['createRole'];
+    inherit: PgRoleConfig['inherit'];
     static readonly [entityKind]: string;
     constructor(name: string, config?: PgRoleConfig);
     existing(): this;
diff --git a/pg-core/roles.d.ts b/pg-core/roles.d.ts
index 443d397..e63aa43 100644
--- a/pg-core/roles.d.ts
+++ b/pg-core/roles.d.ts
@@ -4,8 +4,14 @@ export interface PgRoleConfig {
     createRole?: boolean;
     inherit?: boolean;
 }
-export declare class PgRole implements PgRoleConfig {
+type PgRoleState = {
+    [K in keyof PgRoleConfig]-?: PgRoleConfig[K] | undefined;
+};
+export declare class PgRole implements PgRoleState {
     readonly name: string;
+    createDb: PgRoleConfig['createDb'];
+    createRole: PgRoleConfig['createRole'];
+    inherit: PgRoleConfig['inherit'];
     static readonly [entityKind]: string;
     constructor(name: string, config?: PgRoleConfig);
     existing(): this;
diff --git a/postgres-js/session.d.cts b/postgres-js/session.d.cts
index 3d399b1..b761957 100644
--- a/postgres-js/session.d.cts
+++ b/postgres-js/session.d.cts
@@ -20,7 +20,7 @@ export declare class PostgresJsPreparedQuery<T extends PreparedQueryConfig> exte
     private _isResponseInArrayMode;
     private customResultMapper?;
     static readonly [entityKind]: string;
-    constructor(client: Sql, queryString: string, params: unknown[], logger: Logger, cache: Cache, queryMetadata: {
+    constructor(client: Sql | TransactionSql, queryString: string, params: unknown[], logger: Logger, cache: Cache, queryMetadata: {
         type: 'select' | 'update' | 'delete' | 'insert';
         tables: string[];
     } | undefined, cacheConfig: WithCacheConfig | undefined, fields: SelectedFieldsOrdered | undefined, _isResponseInArrayMode: boolean, customResultMapper?: ((rows: unknown[][]) => T["execute"]) | undefined);
@@ -48,11 +48,14 @@ export declare class PostgresJsSession<TSQL extends Sql, TFullSchema extends Rec
     queryObjects<T extends Row>(query: string, params: unknown[]): Promise<RowList<T[]>>;
     transaction<T>(transaction: (tx: PostgresJsTransaction<TFullSchema, TSchema>) => Promise<T>, config?: PgTransactionConfig): Promise<T>;
 }
+export type PostgresJsTransactionSession<TFullSchema extends Record<string, unknown>, TSchema extends TablesRelationalConfig> = Omit<PostgresJsSession<Sql, TFullSchema, TSchema>, 'client' | 'transaction'> & {
+    client: TransactionSql;
+};
 export declare class PostgresJsTransaction<TFullSchema extends Record<string, unknown>, TSchema extends TablesRelationalConfig> extends PgTransaction<PostgresJsQueryResultHKT, TFullSchema, TSchema> {
     static readonly [entityKind]: string;
     constructor(dialect: PgDialect,
     /** @internal */
-    session: PostgresJsSession<TransactionSql, TFullSchema, TSchema>, schema: RelationalSchemaConfig<TSchema> | undefined, nestedIndex?: number);
+    session: PostgresJsTransactionSession<TFullSchema, TSchema>, schema: RelationalSchemaConfig<TSchema> | undefined, nestedIndex?: number);
     transaction<T>(transaction: (tx: PostgresJsTransaction<TFullSchema, TSchema>) => Promise<T>): Promise<T>;
 }
 export interface PostgresJsQueryResultHKT extends PgQueryResultHKT {
diff --git a/postgres-js/session.d.ts b/postgres-js/session.d.ts
index bccaa53..abfa560 100644
--- a/postgres-js/session.d.ts
+++ b/postgres-js/session.d.ts
@@ -20,7 +20,7 @@ export declare class PostgresJsPreparedQuery<T extends PreparedQueryConfig> exte
     private _isResponseInArrayMode;
     private customResultMapper?;
     static readonly [entityKind]: string;
-    constructor(client: Sql, queryString: string, params: unknown[], logger: Logger, cache: Cache, queryMetadata: {
+    constructor(client: Sql | TransactionSql, queryString: string, params: unknown[], logger: Logger, cache: Cache, queryMetadata: {
         type: 'select' | 'update' | 'delete' | 'insert';
         tables: string[];
     } | undefined, cacheConfig: WithCacheConfig | undefined, fields: SelectedFieldsOrdered | undefined, _isResponseInArrayMode: boolean, customResultMapper?: ((rows: unknown[][]) => T["execute"]) | undefined);
@@ -48,11 +48,14 @@ export declare class PostgresJsSession<TSQL extends Sql, TFullSchema extends Rec
     queryObjects<T extends Row>(query: string, params: unknown[]): Promise<RowList<T[]>>;
     transaction<T>(transaction: (tx: PostgresJsTransaction<TFullSchema, TSchema>) => Promise<T>, config?: PgTransactionConfig): Promise<T>;
 }
+export type PostgresJsTransactionSession<TFullSchema extends Record<string, unknown>, TSchema extends TablesRelationalConfig> = Omit<PostgresJsSession<Sql, TFullSchema, TSchema>, 'client' | 'transaction'> & {
+    client: TransactionSql;
+};
 export declare class PostgresJsTransaction<TFullSchema extends Record<string, unknown>, TSchema extends TablesRelationalConfig> extends PgTransaction<PostgresJsQueryResultHKT, TFullSchema, TSchema> {
     static readonly [entityKind]: string;
     constructor(dialect: PgDialect,
     /** @internal */
-    session: PostgresJsSession<TransactionSql, TFullSchema, TSchema>, schema: RelationalSchemaConfig<TSchema> | undefined, nestedIndex?: number);
+    session: PostgresJsTransactionSession<TFullSchema, TSchema>, schema: RelationalSchemaConfig<TSchema> | undefined, nestedIndex?: number);
     transaction<T>(transaction: (tx: PostgresJsTransaction<TFullSchema, TSchema>) => Promise<T>): Promise<T>;
 }
 export interface PostgresJsQueryResultHKT extends PgQueryResultHKT {
diff --git a/singlestore-core/columns/common.d.cts b/singlestore-core/columns/common.d.cts
index 0dbc570..a6eb254 100644
--- a/singlestore-core/columns/common.d.cts
+++ b/singlestore-core/columns/common.d.cts
@@ -1,5 +1,7 @@
 import type { ColumnBuilderBase, ColumnBuilderBaseConfig, ColumnBuilderExtraConfig, ColumnBuilderRuntimeConfig, ColumnDataType, HasDefault, IsAutoincrement } from "../../column-builder.cjs";
 import { ColumnBuilder } from "../../column-builder.cjs";
+import type { GeneratedColumnConfig, HasGenerated } from "../../column-builder.cjs";
+import type { SQL } from "../../sql/sql.cjs";
 import type { ColumnBaseConfig } from "../../column.cjs";
 import { Column } from "../../column.cjs";
 import { entityKind } from "../../entity.cjs";
@@ -19,6 +21,9 @@ export declare abstract class SingleStoreColumnBuilder<T extends ColumnBuilderBa
 }, TExtraConfig> implements SingleStoreColumnBuilderBase<T, TTypeConfig> {
     static readonly [entityKind]: string;
     unique(name?: string): this;
+    generatedAlwaysAs(as: SQL | T['data'] | (() => SQL), config?: Partial<GeneratedColumnConfig<unknown>>): HasGenerated<this, {
+        type: 'always';
+    }>;
 }
 export declare abstract class SingleStoreColumn<T extends ColumnBaseConfig<ColumnDataType, string> = ColumnBaseConfig<ColumnDataType, string>, TRuntimeConfig extends object = {}, TTypeConfig extends object = {}> extends Column<T, TRuntimeConfig, TTypeConfig & {
     dialect: 'singlestore';
diff --git a/singlestore-core/columns/common.d.ts b/singlestore-core/columns/common.d.ts
index 7d53fe4..6691a48 100644
--- a/singlestore-core/columns/common.d.ts
+++ b/singlestore-core/columns/common.d.ts
@@ -1,5 +1,7 @@
 import type { ColumnBuilderBase, ColumnBuilderBaseConfig, ColumnBuilderExtraConfig, ColumnBuilderRuntimeConfig, ColumnDataType, HasDefault, IsAutoincrement } from "../../column-builder.js";
 import { ColumnBuilder } from "../../column-builder.js";
+import type { GeneratedColumnConfig, HasGenerated } from "../../column-builder.js";
+import type { SQL } from "../../sql/sql.js";
 import type { ColumnBaseConfig } from "../../column.js";
 import { Column } from "../../column.js";
 import { entityKind } from "../../entity.js";
@@ -19,6 +21,9 @@ export declare abstract class SingleStoreColumnBuilder<T extends ColumnBuilderBa
 }, TExtraConfig> implements SingleStoreColumnBuilderBase<T, TTypeConfig> {
     static readonly [entityKind]: string;
     unique(name?: string): this;
+    generatedAlwaysAs(as: SQL | T['data'] | (() => SQL), config?: Partial<GeneratedColumnConfig<unknown>>): HasGenerated<this, {
+        type: 'always';
+    }>;
 }
 export declare abstract class SingleStoreColumn<T extends ColumnBaseConfig<ColumnDataType, string> = ColumnBaseConfig<ColumnDataType, string>, TRuntimeConfig extends object = {}, TTypeConfig extends object = {}> extends Column<T, TRuntimeConfig, TTypeConfig & {
     dialect: 'singlestore';
diff --git a/singlestore-core/columns/enum.d.cts b/singlestore-core/columns/enum.d.cts
index ecd9745..7f76022 100644
--- a/singlestore-core/columns/enum.d.cts
+++ b/singlestore-core/columns/enum.d.cts
@@ -1,4 +1,4 @@
-import type { ColumnBuilderBaseConfig, GeneratedColumnConfig, HasGenerated } from "../../column-builder.cjs";
+import type { ColumnBuilderBaseConfig, GeneratedColumnConfig } from "../../column-builder.cjs";
 import type { ColumnBaseConfig } from "../../column.cjs";
 import { entityKind } from "../../entity.cjs";
 import type { SQL } from "../../sql/index.cjs";
@@ -16,7 +16,7 @@ export type SingleStoreEnumColumnBuilderInitial<TName extends string, TEnum exte
 export declare class SingleStoreEnumColumnBuilder<T extends ColumnBuilderBaseConfig<'string', 'SingleStoreEnumColumn'>> extends SingleStoreColumnBuilder<T, {
     enumValues: T['enumValues'];
 }> {
-    generatedAlwaysAs(as: SQL<unknown> | (() => SQL) | T['data'], config?: Partial<GeneratedColumnConfig<unknown>>): HasGenerated<this, {}>;
+    generatedAlwaysAs(as: SQL<unknown> | (() => SQL) | T['data'], config?: Partial<GeneratedColumnConfig<unknown>>): never;
     static readonly [entityKind]: string;
     constructor(name: T['name'], values: T['enumValues']);
 }
diff --git a/singlestore-core/columns/enum.d.ts b/singlestore-core/columns/enum.d.ts
index a44d79d..e719d5f 100644
--- a/singlestore-core/columns/enum.d.ts
+++ b/singlestore-core/columns/enum.d.ts
@@ -1,4 +1,4 @@
-import type { ColumnBuilderBaseConfig, GeneratedColumnConfig, HasGenerated } from "../../column-builder.js";
+import type { ColumnBuilderBaseConfig, GeneratedColumnConfig } from "../../column-builder.js";
 import type { ColumnBaseConfig } from "../../column.js";
 import { entityKind } from "../../entity.js";
 import type { SQL } from "../../sql/index.js";
@@ -16,7 +16,7 @@ export type SingleStoreEnumColumnBuilderInitial<TName extends string, TEnum exte
 export declare class SingleStoreEnumColumnBuilder<T extends ColumnBuilderBaseConfig<'string', 'SingleStoreEnumColumn'>> extends SingleStoreColumnBuilder<T, {
     enumValues: T['enumValues'];
 }> {
-    generatedAlwaysAs(as: SQL<unknown> | (() => SQL) | T['data'], config?: Partial<GeneratedColumnConfig<unknown>>): HasGenerated<this, {}>;
+    generatedAlwaysAs(as: SQL<unknown> | (() => SQL) | T['data'], config?: Partial<GeneratedColumnConfig<unknown>>): never;
     static readonly [entityKind]: string;
     constructor(name: T['name'], values: T['enumValues']);
 }
diff --git a/singlestore-core/query-builders/delete.d.cts b/singlestore-core/query-builders/delete.d.cts
index 1bc7543..313c8a7 100644
--- a/singlestore-core/query-builders/delete.d.cts
+++ b/singlestore-core/query-builders/delete.d.cts
@@ -73,6 +73,7 @@ export declare class SingleStoreDeleteBase<TTable extends SingleStoreTable, TQue
     orderBy(builder: (deleteTable: TTable) => ValueOrArray<SingleStoreColumn | SQL | SQL.Aliased>): SingleStoreDeleteWithout<this, TDynamic, 'orderBy'>;
     orderBy(...columns: (SingleStoreColumn | SQL | SQL.Aliased)[]): SingleStoreDeleteWithout<this, TDynamic, 'orderBy'>;
     limit(limit: number | Placeholder): SingleStoreDeleteWithout<this, TDynamic, 'limit'>;
+    getSQL(): SQL;
     toSQL(): Query;
     prepare(): SingleStoreDeletePrepare<this>;
     execute: ReturnType<this['prepare']>['execute'];
diff --git a/singlestore-core/query-builders/delete.d.ts b/singlestore-core/query-builders/delete.d.ts
index 64c862c..f655d0c 100644
--- a/singlestore-core/query-builders/delete.d.ts
+++ b/singlestore-core/query-builders/delete.d.ts
@@ -73,6 +73,7 @@ export declare class SingleStoreDeleteBase<TTable extends SingleStoreTable, TQue
     orderBy(builder: (deleteTable: TTable) => ValueOrArray<SingleStoreColumn | SQL | SQL.Aliased>): SingleStoreDeleteWithout<this, TDynamic, 'orderBy'>;
     orderBy(...columns: (SingleStoreColumn | SQL | SQL.Aliased)[]): SingleStoreDeleteWithout<this, TDynamic, 'orderBy'>;
     limit(limit: number | Placeholder): SingleStoreDeleteWithout<this, TDynamic, 'limit'>;
+    getSQL(): SQL;
     toSQL(): Query;
     prepare(): SingleStoreDeletePrepare<this>;
     execute: ReturnType<this['prepare']>['execute'];
diff --git a/singlestore-core/query-builders/select.d.cts b/singlestore-core/query-builders/select.d.cts
index 30fcadc..a986738 100644
--- a/singlestore-core/query-builders/select.d.cts
+++ b/singlestore-core/query-builders/select.d.cts
@@ -511,6 +511,7 @@ export declare abstract class SingleStoreSelectQueryBuilderBase<THKT extends Sin
      * @param config the lock configuration.
      */
     for(strength: LockStrength, config?: LockConfig): SingleStoreSelectWithout<this, TDynamic, 'for'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): SingleStoreSelectDynamic<this>;
diff --git a/singlestore-core/query-builders/select.d.ts b/singlestore-core/query-builders/select.d.ts
index a015f4b..eec58a3 100644
--- a/singlestore-core/query-builders/select.d.ts
+++ b/singlestore-core/query-builders/select.d.ts
@@ -511,6 +511,7 @@ export declare abstract class SingleStoreSelectQueryBuilderBase<THKT extends Sin
      * @param config the lock configuration.
      */
     for(strength: LockStrength, config?: LockConfig): SingleStoreSelectWithout<this, TDynamic, 'for'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): SingleStoreSelectDynamic<this>;
diff --git a/singlestore-core/query-builders/select.types.d.cts b/singlestore-core/query-builders/select.types.d.cts
index 12a5208..c78b9e2 100644
--- a/singlestore-core/query-builders/select.types.d.cts
+++ b/singlestore-core/query-builders/select.types.d.cts
@@ -94,7 +94,7 @@ export interface SingleStoreSelectHKT extends SingleStoreSelectHKTBase {
     _type: SingleStoreSelectBase<this['tableName'], Assume<this['selection'], ColumnsSelection>, this['selectMode'], Assume<this['preparedQueryHKT'], PreparedQueryHKTBase>, Assume<this['nullabilityMap'], Record<string, JoinNullability>>, this['dynamic'], this['excludedMethods'], Assume<this['result'], any[]>, Assume<this['selectedFields'], ColumnsSelection>>;
 }
 export type SingleStoreSetOperatorExcludedMethods = 'where' | 'having' | 'groupBy' | 'session' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'fullJoin' | 'for';
-export type SingleStoreSelectWithout<T extends AnySingleStoreSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SingleStoreSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type SingleStoreSelectWithout<T extends AnySingleStoreSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | SingleStoreSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SingleStoreSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type SingleStoreSelectPrepare<T extends AnySingleStoreSelect> = PreparedQueryKind<T['_']['preparedQueryHKT'], SingleStorePreparedQueryConfig & {
     execute: T['_']['result'];
     iterator: T['_']['result'][number];
diff --git a/singlestore-core/query-builders/select.types.d.ts b/singlestore-core/query-builders/select.types.d.ts
index 761fa02..33449b5 100644
--- a/singlestore-core/query-builders/select.types.d.ts
+++ b/singlestore-core/query-builders/select.types.d.ts
@@ -94,7 +94,7 @@ export interface SingleStoreSelectHKT extends SingleStoreSelectHKTBase {
     _type: SingleStoreSelectBase<this['tableName'], Assume<this['selection'], ColumnsSelection>, this['selectMode'], Assume<this['preparedQueryHKT'], PreparedQueryHKTBase>, Assume<this['nullabilityMap'], Record<string, JoinNullability>>, this['dynamic'], this['excludedMethods'], Assume<this['result'], any[]>, Assume<this['selectedFields'], ColumnsSelection>>;
 }
 export type SingleStoreSetOperatorExcludedMethods = 'where' | 'having' | 'groupBy' | 'session' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'fullJoin' | 'for';
-export type SingleStoreSelectWithout<T extends AnySingleStoreSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SingleStoreSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type SingleStoreSelectWithout<T extends AnySingleStoreSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | SingleStoreSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SingleStoreSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['selection'], T['_']['selectMode'], T['_']['preparedQueryHKT'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type SingleStoreSelectPrepare<T extends AnySingleStoreSelect> = PreparedQueryKind<T['_']['preparedQueryHKT'], SingleStorePreparedQueryConfig & {
     execute: T['_']['result'];
     iterator: T['_']['result'][number];
diff --git a/sqlite-core/query-builders/query.d.cts b/sqlite-core/query-builders/query.d.cts
index 03c8c8f..b78a6ec 100644
--- a/sqlite-core/query-builders/query.d.cts
+++ b/sqlite-core/query-builders/query.d.cts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.cjs";
 import { QueryPromise } from "../../query-promise.cjs";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.cjs";
 import type { RunnableQuery } from "../../runnable-query.cjs";
-import type { Query, SQLWrapper } from "../../sql/sql.cjs";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.cjs";
 import type { KnownKeysOnly } from "../../utils.cjs";
 import type { SQLiteDialect } from "../dialect.cjs";
 import type { PreparedQueryConfig, SQLitePreparedQuery, SQLiteSession } from "../session.cjs";
@@ -46,6 +46,7 @@ export declare class SQLiteRelationalQuery<TType extends 'sync' | 'async', TResu
         execute: TResult;
     }>;
     private _toSQL;
+    getSQL(): SQL;
     toSQL(): Query;
     execute(): Promise<TResult>;
 }
diff --git a/sqlite-core/query-builders/query.d.ts b/sqlite-core/query-builders/query.d.ts
index 06ea6c0..af16e96 100644
--- a/sqlite-core/query-builders/query.d.ts
+++ b/sqlite-core/query-builders/query.d.ts
@@ -2,7 +2,7 @@ import { entityKind } from "../../entity.js";
 import { QueryPromise } from "../../query-promise.js";
 import { type BuildQueryResult, type DBQueryConfig, type TableRelationalConfig, type TablesRelationalConfig } from "../../relations.js";
 import type { RunnableQuery } from "../../runnable-query.js";
-import type { Query, SQLWrapper } from "../../sql/sql.js";
+import type { Query, SQL, SQLWrapper } from "../../sql/sql.js";
 import type { KnownKeysOnly } from "../../utils.js";
 import type { SQLiteDialect } from "../dialect.js";
 import type { PreparedQueryConfig, SQLitePreparedQuery, SQLiteSession } from "../session.js";
@@ -46,6 +46,7 @@ export declare class SQLiteRelationalQuery<TType extends 'sync' | 'async', TResu
         execute: TResult;
     }>;
     private _toSQL;
+    getSQL(): SQL;
     toSQL(): Query;
     execute(): Promise<TResult>;
 }
diff --git a/sqlite-core/query-builders/select.d.cts b/sqlite-core/query-builders/select.d.cts
index 4155c9a..c339bc8 100644
--- a/sqlite-core/query-builders/select.d.cts
+++ b/sqlite-core/query-builders/select.d.cts
@@ -443,6 +443,7 @@ export declare abstract class SQLiteSelectQueryBuilderBase<THKT extends SQLiteSe
      * ```
      */
     offset(offset: number | Placeholder): SQLiteSelectWithout<this, TDynamic, 'offset'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): SQLiteSelectDynamic<this>;
diff --git a/sqlite-core/query-builders/select.d.ts b/sqlite-core/query-builders/select.d.ts
index 44561db..e76acca 100644
--- a/sqlite-core/query-builders/select.d.ts
+++ b/sqlite-core/query-builders/select.d.ts
@@ -443,6 +443,7 @@ export declare abstract class SQLiteSelectQueryBuilderBase<THKT extends SQLiteSe
      * ```
      */
     offset(offset: number | Placeholder): SQLiteSelectWithout<this, TDynamic, 'offset'>;
+    getSQL(): SQL;
     toSQL(): Query;
     as<TAlias extends string>(alias: TAlias): SubqueryWithSelection<this['_']['selectedFields'], TAlias>;
     $dynamic(): SQLiteSelectDynamic<this>;
diff --git a/sqlite-core/query-builders/select.types.d.cts b/sqlite-core/query-builders/select.types.d.cts
index faf89ea..710786e 100644
--- a/sqlite-core/query-builders/select.types.d.cts
+++ b/sqlite-core/query-builders/select.types.d.cts
@@ -82,7 +82,7 @@ export interface SQLiteSelectHKT extends SQLiteSelectHKTBase {
 }
 export type SQLiteSetOperatorExcludedMethods = 'config' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'fullJoin' | 'where' | 'having' | 'groupBy';
 export type CreateSQLiteSelectFromBuilderMode<TBuilderMode extends 'db' | 'qb', TTableName extends string | undefined, TResultType extends 'sync' | 'async', TRunResult, TSelection extends ColumnsSelection, TSelectMode extends SelectMode> = TBuilderMode extends 'db' ? SQLiteSelectBase<TTableName, TResultType, TRunResult, TSelection, TSelectMode> : SQLiteSelectQueryBuilderBase<SQLiteSelectQueryBuilderHKT, TTableName, TResultType, TRunResult, TSelection, TSelectMode>;
-export type SQLiteSelectWithout<T extends AnySQLiteSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SQLiteSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['resultType'], T['_']['runResult'], T['_']['selection'], T['_']['selectMode'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type SQLiteSelectWithout<T extends AnySQLiteSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | SQLiteSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SQLiteSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['resultType'], T['_']['runResult'], T['_']['selection'], T['_']['selectMode'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type SQLiteSelectExecute<T extends AnySQLiteSelect> = T['_']['result'];
 export type SQLiteSelectPrepare<T extends AnySQLiteSelect> = SQLitePreparedQuery<{
     type: T['_']['resultType'];
diff --git a/sqlite-core/query-builders/select.types.d.ts b/sqlite-core/query-builders/select.types.d.ts
index b0030b7..44125e8 100644
--- a/sqlite-core/query-builders/select.types.d.ts
+++ b/sqlite-core/query-builders/select.types.d.ts
@@ -82,7 +82,7 @@ export interface SQLiteSelectHKT extends SQLiteSelectHKTBase {
 }
 export type SQLiteSetOperatorExcludedMethods = 'config' | 'leftJoin' | 'rightJoin' | 'innerJoin' | 'fullJoin' | 'where' | 'having' | 'groupBy';
 export type CreateSQLiteSelectFromBuilderMode<TBuilderMode extends 'db' | 'qb', TTableName extends string | undefined, TResultType extends 'sync' | 'async', TRunResult, TSelection extends ColumnsSelection, TSelectMode extends SelectMode> = TBuilderMode extends 'db' ? SQLiteSelectBase<TTableName, TResultType, TRunResult, TSelection, TSelectMode> : SQLiteSelectQueryBuilderBase<SQLiteSelectQueryBuilderHKT, TTableName, TResultType, TRunResult, TSelection, TSelectMode>;
-export type SQLiteSelectWithout<T extends AnySQLiteSelectQueryBuilder, TDynamic extends boolean, K extends keyof T & string, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SQLiteSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['resultType'], T['_']['runResult'], T['_']['selection'], T['_']['selectMode'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
+export type SQLiteSelectWithout<T extends AnySQLiteSelectQueryBuilder, TDynamic extends boolean, K extends (keyof T & string) | SQLiteSetOperatorExcludedMethods, TResetExcluded extends boolean = false> = TDynamic extends true ? T : Omit<SQLiteSelectKind<T['_']['hkt'], T['_']['tableName'], T['_']['resultType'], T['_']['runResult'], T['_']['selection'], T['_']['selectMode'], T['_']['nullabilityMap'], TDynamic, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K, T['_']['result'], T['_']['selectedFields']>, TResetExcluded extends true ? K : T['_']['excludedMethods'] | K>;
 export type SQLiteSelectExecute<T extends AnySQLiteSelect> = T['_']['result'];
 export type SQLiteSelectPrepare<T extends AnySQLiteSelect> = SQLitePreparedQuery<{
     type: T['_']['resultType'];
